ext/http/00_serve.js

```diff
@@ -472,8 +473,8 @@ function mapToCallback(context, callback, onError) {
     // Did everything shut down while we were waiting?
     if (context.closed) {
       // We're shutting down, so this status shouldn't make it back to the client but "Service Unavailable" seems appropriate
-      op_http_set_promise_complete(req, 503);
       innerRequest?.close();
+      op_http_set_promise_complete(req, 503);
```

Comment:
Why is the order of these two lines swapped?

Reply:
The 'finalizing' ops convert the external to an Rc using `take_external` which does not bump the refcount, effectively consuming the Rc referenced by the external. Swapping the order here ensures that if such an op were to throw an error `innerRequest.#external` would still be set to null so could not be reused.

The 'non-finalizing' ops use `take_clone` which bumps the refcount allowing the external to be reused.

```js
  close() {
    this.#external = null;
  }
```

```rs
#[op2(fast)]
pub fn op_http_set_promise_complete(external: *const c_void, status: u16) {
  let http =
    // SAFETY: external is deleted before calling this op.
    unsafe { take_external!(external, "op_http_set_promise_complete") };
```

```rs
/// Construct Rc<HttpRecord> from raw external pointer, consuming
/// refcount. You must make sure the external is deleted on the JS side.
macro_rules! take_external {
  ($external:expr, $args:tt) => {{
    assert!(
      !$external.is_null(),
      "HTTP state error: external must not be null"
    );
    let record = Rc::from_raw($external as *const HttpRecord);
    http_trace!(record, $args);
    record
  }};
}
```